### PR TITLE
[Validator] Use Mime component to determine mime type for file validator

### PR DESCRIPTION
**Pull request by pierredup**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

When validating the mime type for a file, the Validator component relies on the `Symfony\Component\HttpFoundation\File\File` class, but if the HttpFoundation component is not installed, then you just get the error

```
PHP Fatal error:  Uncaught Error: Class 'Symfony\Component\HttpFoundation\File\File' not found
```

This PR uses the Mime component to get the mime type for a file and throws an exception if the Mime component is not installed.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Now that the mime component is standalone, can't we skip the dependency on HttpFoundation instead?

**Comment by pierredup** [comment]
@nicolas-grekas Done

File: src/Symfony/Component/Validator/Constraints/FileValidator.php [hunk 1]
```diff
@@ -170,12 +172,12 @@ public function validate($value, Constraint $constraint)
         }
 
         if ($constraint->mimeTypes) {
-            if (!$value instanceof FileObject) {
-                $value = new FileObject($value);
+            if (!class_exists(MimeTypes::class)) {
+                throw new LogicException('Validating mime types requires the "Mime" component. Install "symfony/mime" to use it.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/FileValidator.php:176]: let's reuse the template given by existing messages, e.g.:
`You cannot validate the mime-type of files as the Mime component is not installed. Try running "composer require symfony/mime".`

**Comment by nicolas-grekas** [comment]
What about considering it as a bugfix?

**Comment by pierredup** [comment]
> What about considering it as a bugfix?

I wasn't sure if it counted as a bugfix or new feature. When in doubt, I just consider something a new feature :)

But I'm not sure if a bugfix against 3.4 can use the Mime component, so should I create a new PR for 3.4 and have the original exception message for `HttpFoundation`, then just rebase and retarget 4.4 here?

**Comment by nicolas-grekas** [comment]
I would suggest targeting 4.4 and forget about 3.4.

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/FileValidator.php [hunk 2]
```diff
@@ -170,12 +172,12 @@ public function validate($value, Constraint $constraint)
         }
 
         if ($constraint->mimeTypes) {
-            if (!$value instanceof FileObject) {
-                $value = new FileObject($value);
+            if (!class_exists(MimeTypes::class)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/FileValidator.php:175]: Won't this break applications where people use the Validator component in 4.4, but still use version 3.4 of the HttpFoundation component (previously that combination was able to successfully validate)?

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/FileValidator.php:175]: You're right! @pierredup can you please update to account for the case were only HttpFoundation is found?

**Comment by pierredup** [comment]
Status: Ready for review

**Comment by nicolas-grekas** [comment]
I pushed a CS change. The code looks good, but the tests need some love, see [failures on appveyor](https://ci.appveyor.com/project/fabpot/symfony/builds/33194124) when the finfo extension is not enabled.

**Comment by pierredup** [comment]
I reverted the changes to the tests to use mocks again, and just changes the ordering of the checks

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/FileValidator.php [hunk 3]
```diff
@@ -170,12 +172,17 @@ public function validate($value, Constraint $constraint)
...
 
         if ($constraint->mimeTypes) {
-            if (!$value instanceof FileObject) {
-                $value = new FileObject($value);
+            if ($value instanceof FileObject) {
+                $mime = $value->getMimeType();
+            } elseif (class_exists(MimeTypes::class)) {
+                $mime = MimeTypes::getDefault()->guessMimeType($path);
+            } elseif (!class_exists(FileObject::class)) {
+                throw new LogicException('You cannot validate the mime-type of files as the Mime component is not installed. Try running "composer require symfony/mime".');
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/FileValidator.php:180]: They are correct: the class checked is for supporting HttpFoundation < 4.4, and the suggestion is supposing Symfony >= 4.4

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @pierredup.

**Pull request merged**